In [2]:
import pandas as pd
import numpy as np

In [3]:
from sklearn.preprocessing import OneHotEncoder,LabelEncoder,StandardScaler
from sklearn.model_selection import train_test_split

In [4]:
import pickle

In [5]:
data = pd.read_csv('Churn_Modelling.csv')
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [7]:
data.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [8]:
data.duplicated().sum()

0

In [9]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [10]:
data = data.drop(columns=['RowNumber','CustomerId','Surname'],axis=1)

In [11]:
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [12]:
data['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

In [13]:
data.Gender.unique()

array(['Female', 'Male'], dtype=object)

In [14]:
# creating instance of Label Encoder

l_encoder_gender = LabelEncoder()

In [15]:
data['Gender'] = l_encoder_gender.fit_transform(data.Gender)

In [16]:
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France       0   42       2       0.00              1   
1          608     Spain       0   41       1   83807.86              1   
2          502    France       0   42       8  159660.80              3   
3          699    France       0   39       1       0.00              2   
4          850     Spain       0   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [17]:
# instance for onehotencoder
one_hot_encoder_geo = OneHotEncoder()

In [18]:
geo_encoded = one_hot_encoder_geo.fit_transform(data[['Geography']]).toarray()

In [19]:
one_hot_encoder_geo.get_feature_names_out(['Geography'])

array(['Geography_France', 'Geography_Germany', 'Geography_Spain'],
      dtype=object)

In [20]:
geo_data_df = pd.DataFrame(geo_encoded,columns=one_hot_encoder_geo.get_feature_names_out(['Geography']))
geo_data_df.head()

Geography_France  Geography_Germany  Geography_Spain
0               1.0                0.0              0.0
1               0.0                0.0              1.0
2               1.0                0.0              0.0
3               1.0                0.0              0.0
4               0.0                0.0              1.0

In [21]:
# Combine data with transformed geo data
data = pd.concat([data.drop(columns=['Geography'],axis=1),geo_data_df],axis=1)

In [22]:
data.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_France  \
0               1        101348.88       1               1.0   
1               1        112542.58       0               0.0   
2               0        113931.57       1               1.0   
3               0         93826.63       0               1.0   
4               1         79084.10       0               0.0   

   Geography_Germany  Geography_Spain  
0                0.0              0.0  
1                0.0              1.0  
2                0.0              0.0  
3                0.0              0.0  
4                0.0              1.0

In [23]:
#Saving encoder in pickle file for later use
with open('label_encoder_gender.pkl','wb') as file:
    pickle.dump(l_encoder_gender,file)

with open('one_hot_encoder_geo.pkl','wb') as file:
    pickle.dump(one_hot_encoder_geo,file)

In [24]:
data.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_France  \
0               1        101348.88       1               1.0   
1               1        112542.58       0               0.0   
2               0        113931.57       1               1.0   
3               0         93826.63       0               1.0   
4               1         79084.10       0               0.0   

   Geography_Germany  Geography_Spain  
0                0.0              0.0  
1                0.0              1.0  
2                0.0              0.0  
3                0.0              0.0  
4                0.0              1.0

In [25]:
# Defining independ and depend variable
X = data.drop(columns=['Exited'],axis=1)
y = data['Exited']

In [26]:
X.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619       0   42       2       0.00              1          1   
1          608       0   41       1   83807.86              1          0   
2          502       0   42       8  159660.80              3          1   
3          699       0   39       1       0.00              2          0   
4          850       0   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
0               1        101348.88               1.0                0.0   
1               1        112542.58               0.0                0.0   
2               0        113931.57               1.0                0.0   
3               0         93826.63               1.0                0.0   
4               1         79084.10               0.0                0.0   

   Geography_Spain  
0              0.0  
1              1.0  
2              0.0  
3              0.0  
4              1.0

In [27]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [28]:
# Spliting the dataset into train and test

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [125]:
X_train

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
9254          686       1   32       6       0.00              2          1   
1561          632       1   42       4  119624.60              2          1   
1670          559       1   24       3  114739.92              1          1   
6087          561       0   27       9  135637.00              1          1   
6669          517       1   56       9  142147.32              1          0   
...           ...     ...  ...     ...        ...            ...        ...   
5734          768       1   54       8   69712.74              1          1   
5191          682       0   58       1       0.00              1          1   
5390          735       0   38       1       0.00              3          0   
860           667       1   43       8  190227.46              1          1   
7270          697       1   51       1  147910.30              1          1   

      IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
9254               1        179093.26               1.0                0.0   
1561               1        195978.86               0.0                1.0   
1670               0         85891.02               0.0                0.0   
6087               0        153080.40               1.0                0.0   
6669               0         39488.04               1.0                0.0   
...              ...              ...               ...                ...   
5734               1         69381.05               1.0                0.0   
5191               1           706.50               1.0                0.0   
5390               0         92220.12               1.0                0.0   
860                0         97508.04               1.0                0.0   
7270               1         53581.14               0.0                1.0   

      Geography_Spain  
9254              0.0  
1561              0.0  
1670              1.0  
6087              0.0  
6669              0.0  
...               ...  
5734              0.0  
5191              0.0  
5390              0.0  
860               0.0  
7270              0.0  

[8000 rows x 12 columns]

In [75]:
X_test

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
6252          596       1   32       3   96709.07              2          0   
4684          623       1   43       1       0.00              2          1   
1731          601       0   44       4       0.00              2          1   
4742          506       1   59       8  119152.10              2          1   
4521          560       0   27       7  124995.98              1          1   
...           ...     ...  ...     ...        ...            ...        ...   
6412          602       0   53       5   98268.84              1          0   
8285          609       1   25      10       0.00              1          0   
7853          730       0   47       7       0.00              1          1   
1095          692       1   29       4       0.00              1          1   
6929          628       1   39       1  115341.19              1          1   

      IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
6252               0         41788.37               0.0                1.0   
4684               1        146379.30               1.0                0.0   
1731               0         58561.31               0.0                0.0   
4742               1        170679.74               0.0                1.0   
4521               1        114669.79               0.0                0.0   
...              ...              ...               ...                ...   
6412               1         45038.29               0.0                1.0   
8285               1        109895.16               1.0                0.0   
7853               0         33373.26               1.0                0.0   
1095               0         76755.99               1.0                0.0   
6929               1        107674.30               0.0                1.0   

      Geography_Spain  
6252              0.0  
4684              0.0  
1731              1.0  
4742              0.0  
4521              1.0  
...               ...  
6412              0.0  
8285              0.0  
7853              0.0  
1095              0.0  
6929              0.0  

[2000 rows x 12 columns]

In [35]:
y_train

9254    0
1561    0
1670    1
6087    1
6669    1
       ..
5734    0
5191    0
5390    1
860     1
7270    0
Name: Exited, Length: 8000, dtype: int64

In [29]:
# Scalling
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [30]:
X_train

array([[ 0.35649971,  0.91324755, -0.6557859 , ...,  1.00150113,
        -0.57946723, -0.57638802],
       [-0.20389777,  0.91324755,  0.29493847, ..., -0.99850112,
         1.72572313, -0.57638802],
       [-0.96147213,  0.91324755, -1.41636539, ..., -0.99850112,
        -0.57946723,  1.73494238],
       ...,
       [ 0.86500853, -1.09499335, -0.08535128, ...,  1.00150113,
        -0.57946723, -0.57638802],
       [ 0.15932282,  0.91324755,  0.3900109 , ...,  1.00150113,
        -0.57946723, -0.57638802],
       [ 0.47065475,  0.91324755,  1.15059039, ..., -0.99850112,
         1.72572313, -0.57638802]])

In [31]:
# Saving scaler 
with open('scaler.pkl','wb') as file:
    pickle.dump(scaler,file)

## ANN Implemetation

In [32]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.callbacks import EarlyStopping,TensorBoard

In [33]:
(X_train.shape[1],)

(12,)

In [34]:
# Building our connected layer and ann model
model = Sequential([
    Dense(64,activation = 'relu', input_shape=(X_train.shape[1],)), # HL1 connected with input layer
    Dense(32,activation='relu'),                                    # HL2
    Dense(1,activation='sigmoid')                                   # Output layer

])

In [35]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 64)                832       
                                                                 
 dense_1 (Dense)             (None, 32)                2080      
                                                                 
 dense_2 (Dense)             (None, 1)                 33        
                                                                 
Total params: 2945 (11.50 KB)
Trainable params: 2945 (11.50 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [36]:
# Model Compile (loss,optimize)
import tensorflow
loss = tensorflow.keras.losses.BinaryCrossentropy()
opt = tensorflow.keras.optimizers.Adam(learning_rate=0.01)

In [37]:
model.compile(loss=loss,optimizer=opt,metrics=['accuracy'])

In [38]:
# Setup EArlystopping ---> stop traing when loss decrease is not much
early_stopping_callbacks = EarlyStopping(monitor='val_loss', patience=5,restore_best_weights=True)

In [39]:
# Setup tensorboard
import datetime

log_dir = "log/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorflow_callbacks = TensorBoard(log_dir=log_dir,histogram_freq=1)

In [40]:
history = model.fit(
    X_train,y_train,
    validation_data = (X_test,y_test),
    epochs = 40,
    callbacks = [tensorflow_callbacks,early_stopping_callbacks]
)

Epoch 1/40


250/250 [==============================] - 3s 5ms/step - loss: 0.3938 - accuracy: 0.8349 - val_loss: 0.3657 - val_accuracy: 0.8470
Epoch 2/40
250/250 [==============================] - 1s 4ms/step - loss: 0.3543 - accuracy: 0.8514 - val_loss: 0.3479 - val_accuracy: 0.8565
Epoch 3/40
250/250 [==============================] - 1s 3ms/step - loss: 0.3473 - accuracy: 0.8561 - val_loss: 0.3644 - val_accuracy: 0.8540
Epoch 4/40
250/250 [==============================] - 1s 4ms/step - loss: 0.3420 - accuracy: 0.8608 - val_loss: 0.3419 - val_accuracy: 0.8595
Epoch 5/40
250/250 [==============================] - 1s 3ms/step - loss: 0.3403 - accuracy: 0.8626 - val_loss: 0.3418 - val_accuracy: 0.8590
Epoch 6/40
250/250 [==============================] - 1s 3ms/step - loss: 0.3392 - accuracy: 0.8619 - val_loss: 0.3547 - val_accuracy: 0.8535
Epoch 7/40
250/250 [==============================] - 1s 4ms/step - loss: 0.3342 - accuracy: 0.8648 - val_loss: 0.3492 - val_accuracy: 0.8505
Epoc

In [41]:
model.save('model.h5')

d:\AI\Projects\DL\ANNproject\myvenv\Lib\site-packages\keras\src\engine\training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [98]:
# Load Tensorboard extension
%load_ext tensorboard

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


In [101]:
%tensorboard --logdir log/fit/20250826-113707

Reusing TensorBoard on port 6006 (pid 27916), started 0:00:04 ago. (Use '!kill 27916' to kill it.)

#### To find the running tensorflow
netstat -ano | findstr 6006 0r 6008{#port number}

#### To kill the task from baxkground
taskkill /PID 27916{# process id} /F